Projections

prolic opened the pull request.

prolic commented:
Please carefully check the locking mechanisms!

lucascourot commented on src/Projection/MySQLEventStoreProjectionTrait.php:
```diff
@@ -0,0 +1,62 @@
+     */
+    protected function acquireLock(): void
+    {
+        $statement = $this->connection->prepare("SELECT GET_LOCK('$this->name', 1) as 'lock';");
```
no need to surround aliases with quotes

prolic replied:
It's required!

lucascourot replied:
why? can't you do simply ` as lock`?

prolic replied:
No.

```
mysql> SELECT get_lock('ff', 3) as lock;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'lock' at line 1
```

lucascourot replied:
oh ok, fair enough, the 'lock' keyword is reserved :)

prolic replied:
damn, that's a bug. or maybe we do the same implementation as in psql?

prolic replied:
I think having the same implementation in place might be the best scenario. Will update the code tomorrow.

lucascourot replied:
Ok but don't forget to expose the timeout value in the config of the package so the user can configure it. I would actually be in favor of letting the user chose how he really wants to manage his locks by exposing a locking interface. Another solution would be to make your locked column a datetime meaning until when the projection is locked (instead of storing a random int).

lucascourot commented on src/Projection/PostgresEventStoreProjectionTrait.php:
```diff
@@ -0,0 +1,95 @@
+//die;
+        if (false === $row) {
+            $sql = <<<EOT
+INSERT INTO $this->projectionsTable (name, position, state, locked) 
```
so you use GET_LOCK() for MySQL but a locked column for Postgres?

prolic replied:
Yes. PostgreSQL has not the same feature.

prolic commented on src/Projection/PostgresEventStoreProjectionTrait.php:
```diff
@@ -0,0 +1,95 @@
+        $statement->execute();
+
+        $row = $statement->fetch(PDO::FETCH_OBJ);
+//die;
```
done

prolic commented:
damn, that's a bug. or maybe we do the same implementation as in psql?

Sascha-Oliver Prolic

2016-11-17 0:25 GMT+08:00 Lucas Courot <notifications@github.com>:

> *@lucascourot* commented on this pull request.
> ------------------------------
>
> In src/Projection/MySQLEventStoreProjectionTrait.php
> <https://github.com/prooph/pdo-event-store/pull/9>:
>
> > +
> +declare(strict_types=1);
> +
> +namespace Prooph\EventStore\PDO\Projection;
> +
> +use PDO;
> +use Prooph\EventStore\PDO\Exception\RuntimeException;
> +
> +trait MySQLEventStoreProjectionTrait
> +{
> +    /**
> +     * @throws RuntimeException
> +     */
> +    protected function acquireLock(): void
> +    {
> +        $statement = $this->connection->prepare("SELECT GET_LOCK('$this->name', 1) as 'lock';");
>
> btw, I see you pass 1 as a second parameter of GET_LOCK, which means your
> lock will be released after 1 second. This is not what we want, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prooph/pdo-event-store/pull/9>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAYEvELblWSvDaidtg9dE0ESYqSjF4naks5q-y51gaJpZM4KzhRs>
> .
>

prolic commented:
There is also a pthreads implementation planned for multi stream processing. Problem is different threads cannot share db connection. So how to construct this object? Inject a container and service names? Any ideas?

codeliner commented:
> There is also a pthreads implementation planned for multi stream processing. Problem is different threads cannot share db connection.

I followed @bendavies suggestion in the chat and took a look at icicle. I played a bit with reactPHP in the past and also read a few things about icicle.
Most of the examples are dealing with non-blocking streams but a google search took me there: https://packagist.org/packages/asyncphp/icicle-database
Except my experiments at home, I did not use non-blocking techniques in php so I cannot say much about it. Maybe we can ask the async php community for help or ideas ... non-blocking I/O works great for node.js and Go so why not also for php, but database connections would need to support it.

prolic commented:
- removed deadlocks
- added lock until timeout in milliseconds
- added projection factories
- added query factories
- added read model projection factories

bendavies commented on src/Projection/PDOEventStoreProjectionTrait.php:
```diff
@@ -0,0 +1,76 @@
+        $lockUntilString = $now->modify('+' . (string) $this->lockTimeoutMs . ' ms')->format('Y-m-d\TH:i:s.u');
+
+        $sql = <<<EOT
+UPDATE $this->projectionsTable SET locked_until = ? WHERE name = ? AND (locked_until IS NuLL OR locked_until < ?);
```
`NULL`

prolic replied:
done

prolic commented:
About multi-stream-processing:

This will require pthreads and will probably be added with in a distinct implementation (much stuff differs then). Still will be added later and is not required at the moment.

prolic commented:
Note: This PR is now also cleaned up, so a DistributedLock interface can be added very easy (but is not yet done in this PR).

prolic commented:
Travis with postgres 9.4 hangs. But should be no problem, it's just travis :)

codeliner commented on src/Projection/MySQLEventStoreProjection.php:
```diff
@@ -0,0 +1,77 @@
+        $this->connection->beginTransaction();
+
+        $listener = $this->eventStore->getActionEventEmitter()->attachListener(
+            ActionEventEmitterAware::EVENT_APPEND_TO,
```
wrong event?

prolic replied:
fixed

codeliner commented on src/Projection/MySQLEventStoreReadModelProjection.php:
```diff
@@ -0,0 +1,80 @@
+        $this->connection->beginTransaction();
+
+        $listener = $this->eventStore->getActionEventEmitter()->attachListener(
+            ActionEventEmitterAware::EVENT_APPEND_TO,
```
wrong event?

prolic replied:
removed
